### ISSUE TITLE
[Feature Request] Add button to load a custom workflow

### ISSUE THREAD

**Issue by ljleb** [ModelSurge/sd-webui-comfyui#113]
Add a dropdown listing workflow types and a button to load the current selection. The extension would discover workflow types from a list of paths in the settings and when clicking on a reload button.

**Comment by ljleb** [comment]
Another option that would come in handy is a way to customize the default workflow to load for individual workflow types.

**Comment by toyxyz** [comment]
This is useful for saving and loading custom workflows! https://github.com/pythongosssss/ComfyUI-Custom-Scripts

**Comment by John-WL** [comment]
This sounds exactly like the same feature, indeed. Thank you for sharing. The problem is that it adds the dropdown in the UI menu, which is hidden to improve visibility and prevent some features in the ComfyUI clients under the txt2img and img2img tabs. 

I think this means that we won't have to implement it in the "sandbox" tab, though, which is nice. 

We might still need to implement a gradio dropdown in the UI for the txt2img and img2img tabs.

**Comment by ljleb** [comment]
related: #57

**Comment by ljleb** [comment]
I think it would make sense to distinguish the type of workflows when scanning them, so that for example one can filter latent IO workflows. Workflow types have to have a specific type of data as input and output. It should be obvious whether a loaded workflow would run with the type of the input and output nodes without running a generation.

Before (or while) implementing this, we may want to consider displaying an error or some nodes in red, if the workflow is detected to be using invalid input and output types.

**Comment by John-WL** [comment]
Yes, I think there is a way to implement this with simple S&R name validation.

**Comment by ljleb** [comment]
This is true, if we keep maintaining 2 nodes (from and to webui) per type of data. If we change this so that the same node is used for every type, I'm not sure we can rely on S&R validation. We could pass the type of data as an input to the nodes maybe? These serialized type names would end up in the workflow, so we could use this information for validation.

**Comment by John-WL** [comment]
We could also add use the js extension to add additional info in Webui nodes if needed.

**Comment by 4lt3r3go** [comment]
> This is useful for saving and loading custom workflows! https://github.com/pythongosssss/ComfyUI-Custom-Scripts

i'm trying everything including this. unable to load back the workflow, it load the default one once i import.
i wonder if there's a way to save the workflow too, as reported here https://github.com/ModelSurge/sd-webui-comfyui/issues/219
the Ui menu of comfy disappear after few seconds here, resulting in a complete grey board wich loads module but cant access to the menu

**Comment by toyxyz** [comment]
https://github.com/Nuked88/ComfyUI-N-Sidebar
![workflows](https://github.com/ModelSurge/sd-webui-comfyui/assets/8006000/fcfd6ef6-d5dc-4a05-af49-1899a8ad3b56)

**Comment by 4lt3r3go** [comment]
> https://github.com/Nuked88/ComfyUI-N-Sidebar [![workflows](https://private-user-images.githubusercontent.com/8006000/334826907-fcfd6ef6-d5dc-4a05-af49-1899a8ad3b56.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5OTY4ODEsIm5iZiI6MTcxNjk5NjU4MSwicGF0aCI6Ii84MDA2MDAwLzMzNDgyNjkwNy1mY2ZkNmVmNi1kNWRjLTRhMDUtYWY0OS0xODk5YThhZDNiNTYuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MjlUMTUyOTQxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmI2OTExNzEzOWUwZDI5NDJkNDIzYjNjOTU0YTFkMTI5MTFkNjUzMGY0MjE2N2ZmM2E1YTE3MDFmZjYwN2Q0NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.43ykGgLnJela2GjTUOflgRWKsBQ01-nSTVjpih8sjhE)](https://private-user-images.githubusercontent.com/8006000/334826907-fcfd6ef6-d5dc-4a05-af49-1899a8ad3b56.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5OTY4ODEsIm5iZiI6MTcxNjk5NjU4MSwicGF0aCI6Ii84MDA2MDAwLzMzNDgyNjkwNy1mY2ZkNmVmNi1kNWRjLTRhMDUtYWY0OS0xODk5YThhZDNiNTYuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MjlUMTUyOTQxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmI2OTExNzEzOWUwZDI5NDJkNDIzYjNjOTU0YTFkMTI5MTFkNjUzMGY0MjE2N2ZmM2E1YTE3MDFmZjYwN2Q0NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.43ykGgLnJela2GjTUOflgRWKsBQ01-nSTVjpih8sjhE)

added this ty but is not showing anything

**Comment by toyxyz** [comment]
> > https://github.com/Nuked88/ComfyUI-N-Sidebar ![workflows](https://private-user-images.githubusercontent.com/8006000/334826907-fcfd6ef6-d5dc-4a05-af49-1899a8ad3b56.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5OTY4ODEsIm5iZiI6MTcxNjk5NjU4MSwicGF0aCI6Ii84MDA2MDAwLzMzNDgyNjkwNy1mY2ZkNmVmNi1kNWRjLTRhMDUtYWY0OS0xODk5YThhZDNiNTYuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MjlUMTUyOTQxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmI2OTExNzEzOWUwZDI5NDJkNDIzYjNjOTU0YTFkMTI5MTFkNjUzMGY0MjE2N2ZmM2E1YTE3MDFmZjYwN2Q0NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.43ykGgLnJela2GjTUOflgRWKsBQ01-nSTVjpih8sjhE)
> >       
> >         [
> >           
> >             ![workflows](https://private-user-images.githubusercontent.com/8006000/334826907-fcfd6ef6-d5dc-4a05-af49-1899a8ad3b56.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5OTY4ODEsIm5iZiI6MTcxNjk5NjU4MSwicGF0aCI6Ii84MDA2MDAwLzMzNDgyNjkwNy1mY2ZkNmVmNi1kNWRjLTRhMDUtYWY0OS0xODk5YThhZDNiNTYuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MjlUMTUyOTQxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmI2OTExNzEzOWUwZDI5NDJkNDIzYjNjOTU0YTFkMTI5MTFkNjUzMGY0MjE2N2ZmM2E1YTE3MDFmZjYwN2Q0NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.43ykGgLnJela2GjTUOflgRWKsBQ01-nSTVjpih8sjhE)
> >           
> >         ](https://private-user-images.githubusercontent.com/8006000/334826907-fcfd6ef6-d5dc-4a05-af49-1899a8ad3b56.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5OTY4ODEsIm5iZiI6MTcxNjk5NjU4MSwicGF0aCI6Ii84MDA2MDAwLzMzNDgyNjkwNy1mY2ZkNmVmNi1kNWRjLTRhMDUtYWY0OS0xODk5YThhZDNiNTYuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MjlUMTUyOTQxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmI2OTExNzEzOWUwZDI5NDJkNDIzYjNjOTU0YTFkMTI5MTFkNjUzMGY0MjE2N2ZmM2E1YTE3MDFmZjYwN2Q0NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.43ykGgLnJela2GjTUOflgRWKsBQ01-nSTVjpih8sjhE)
> >         
> >         
> >           
> >             
> >               
> >             
> >             
> >               
> >               
> >             
> >           
> >           [
> >             
> >               
> >             
> >           ](https://private-user-images.githubusercontent.com/8006000/334826907-fcfd6ef6-d5dc-4a05-af49-1899a8ad3b56.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5OTY4ODEsIm5iZiI6MTcxNjk5NjU4MSwicGF0aCI6Ii84MDA2MDAwLzMzNDgyNjkwNy1mY2ZkNmVmNi1kNWRjLTRhMDUtYWY0OS0xODk5YThhZDNiNTYuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MjlUMTUyOTQxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmI2OTExNzEzOWUwZDI5NDJkNDIzYjNjOTU0YTFkMTI5MTFkNjUzMGY0MjE2N2ZmM2E1YTE3MDFmZjYwN2Q0NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.43ykGgLnJela2GjTUOflgRWKsBQ01-nSTVjpih8sjhE)
> 
> added this ty but is not showing anything

You must enter a workflow path in the settings of sidebar.

**Comment by toyxyz** [comment]
You must enter a workflow path in the settings of sidebar.
